nexus/src/app/background/tasks/blueprint_rendezvous.rs

```diff
@@ -0,0 +1,115 @@
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::DataStore;
+use nexus_reconfigurator_rendezvous::reconcile_blueprint_rendezvous_tables;
+use nexus_types::deployment::{Blueprint, BlueprintTarget};
+use serde_json::json;
+use slog_error_chain::InlineErrorChain;
+use std::sync::Arc;
+use tokio::sync::watch;
+
+/// Background task that takes a [`Blueprint`] and an inventory [`Collection`]
+/// and updates any rendezvous tables to track resources under Reconfigurator's
+/// control for other parts of Nexus to consume.
+pub struct BlueprintRendezvous {
+    datastore: Arc<DataStore>,
+    rx_blueprint: watch::Receiver<Option<Arc<(BlueprintTarget, Blueprint)>>>,
+}
+
+impl BlueprintRendezvous {
+    pub fn new(
+        datastore: Arc<DataStore>,
+        rx_blueprint: watch::Receiver<
+            Option<Arc<(BlueprintTarget, Blueprint)>>,
+        >,
+    ) -> Self {
+        Self { datastore, rx_blueprint }
+    }
+
+    /// Implementation for `BackgroundTask::activate` for `BlueprintRendezvous`,
+    /// added here to produce better compile errors.
+    ///
+    /// The presence of `boxed()` in `BackgroundTask::activate` has caused some
+    /// confusion with compilation errors in the past. So separate this method
+    /// out.
+    async fn activate_impl<'a>(
+        &mut self,
+        opctx: &OpContext,
+    ) -> serde_json::Value {
+        // Get the latest blueprint, cloning to prevent holding a read lock
+        // on the watch.
+        let update = self.rx_blueprint.borrow_and_update().clone();
+
+        let Some(update) = update else {
+            warn!(
+                &opctx.log, "Blueprint rendezvous: skipped";
+                "reason" => "no blueprint",
+            );
+            return json!({"error": "no blueprint" });
+        };
+
+        // Get the latest inventory collection
+        let maybe_collection = match self
+            .datastore
+            .inventory_get_latest_collection(opctx)
+            .await
+        {
+            Ok(maybe_collection) => maybe_collection,
+            Err(err) => {
+                let err = InlineErrorChain::new(&err);
+                warn!(
+                    &opctx.log, "Blueprint rendezvous: skipped";
+                    "reason" => "failed to read latest inventory collection",
+                    &err,
+                );
+                return json!({ "error":
+                    format!("failed reading inventory collection: {err}"),
+                });
+            }
+        };
+
+        let Some(collection) = maybe_collection else {
+            warn!(
+                &opctx.log, "Blueprint rendezvous: skipped";
+                "reason" => "no inventory collection",
+            );
+            return json!({"error": "no inventory collection" });
+        };
+
+        // Actually perform rendezvous table reconciliation
+        let (_, blueprint) = &*update;
```

Comment:
Nit: Would it be a problem to do this de-structuring above? I think the name `update` is a little confusing

Reply:
Nope and agreed; fixed in [7b5bf53](https://github.com/oxidecomputer/omicron/pull/7342/commits/7b5bf5364eb1389a81baf8f0f7f1ceb93611cd53)